src/lib.rs

```diff
@@ -862,6 +939,46 @@ impl List {
         }
     }
 
+    fn notify_with_token(&mut self, token: T) {
+        if self.notified > 0 {
+            return;
+        }
+
+        self.notify_internal(false, token);
+    }
+
+    fn notify_additional_with_token(&mut self, token: T) {
+        self.notify_internal(true, token);
+    }
+
+    #[inline(always)]
+    fn notify_internal(&mut self, additional: bool, token: T) -> bool {
```

Comment:
I separated the logic for notification on `List` out into this single function which all the notification functions on List call into.